packages/react-reconciler/src/ReactFiberBeginWork.new.js

```diff
@@ -1560,17 +1748,68 @@ const SUSPENDED_MARKER: SuspenseState = {
 };
 
 function mountSuspenseOffscreenState(renderLanes: Lanes): OffscreenState {
+  let cache = null;
+  if (enableCache) {
+    // Keep a reference to the in-flight cache so we can resume later.
+    cache = getFreshCacheProviderIfExists();
+    if (cache === null) {
+      // If there's no cache on the stack, a nested Cache boundary may have
+      // spawned a new one. Check the cache pool.
+      const root = getWorkInProgressRoot();
+      invariant(
+        root !== null,
+        'Expected a work-in-progress root. This is a bug in React. Please ' +
+          'file an issue.',
+      );
+      // If a nested cache accessed the pool during this render, it will be
+      // assigned to root.pooledCache. No need to check the lane-indexed pool.
+      // TODO: Actually I think I'm wrong and we do need to check the lane-indexed
+      // pool, to account for infinite transitions that are not triggered by a
+      // `refresh` call, since those won't put a fresh context on the stack.
+      // However, that's not idiomatic so this might be fine for now.
+      cache = root.pooledCache;
+    }
+  }
   return {
     baseLanes: renderLanes,
+    cache,
   };
 }
 
 function updateSuspenseOffscreenState(
   prevOffscreenState: OffscreenState,
   renderLanes: Lanes,
 ): OffscreenState {
+  let cache = null;
+  if (enableCache) {
+    // Keep a reference to the in-flight cache so we can resume later.
+    cache = getFreshCacheProviderIfExists();
+    if (cache === null) {
+      // If there's no cache on the stack, a nested Cache boundary may have
+      // spawned a new one. Check the cache pool.
+      const root = getWorkInProgressRoot();
+      invariant(
+        root !== null,
+        'Expected a work-in-progress root. This is a bug in React. Please ' +
+          'file an issue.',
+      );
+      // If a nested cache accessed the pool during this render, it will be
+      // assigned to root.pooledCache. No need to check the lane-indexed pool.
+      // TODO: Actually I think I'm wrong and we do need to check the lane-indexed
+      // pool, to account for infinite transitions that are not triggered by a
+      // `refresh` call, since those won't put a fresh context on the stack.
+      // However, that's not idiomatic so this might be fine for now.
+      cache = root.pooledCache;
+      if (cache === null) {
+        // If there's no cache in the pool, there might be one from a previous
+        // render. If so, reuse it.
+        cache = prevOffscreenState.cache;
```

Comment:
Why is this after the pool check? Just because a sibling accessed the pool, why would that affect this component?

Reply:
This function is confusingly named. In this path, we're about to commit a Suspense fallback. So this logic is about deciding whether and which cache we should stash on the boundary so it can be resumed later.

Checking the pool cache is necessary because there may have been a fresh cache in the suspended tree.

Reply:
Oh I misread your question.

Yeah I think you're right. If this is an update, then there's no reason the child should have accessed the pool cache. It would have accessed this boundary's cache. I'll fix this!

(I'll write tests for all these edge cases, btw.)